### PR TITLE
#146 Terms of Service Phase 2

### PR DESCRIPTION
**Pull request by harshilsharma**
#### Summary
1. Added new table `UserTermsOfService` for storing user's accepted terms of service details.
1. Removed model field `User.AcceptedTermsOfServiceId`.
1. Removed column `AcceptedTermsOfServiceId` from `Users` table.
1. Added migration script to migrate user's accepted terms of service data to this new table.
1. Added `UpgradeDatabaseToVersion56()` method in `upgrade.go` and added corresponding version variables.
1. Added APIs for fetching user's TOS details.
1. Added config setting for 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-12400

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs) - https://github.com/mattermost/mattermost-api-reference/pull/402
- [x] All new/modified APIs include changes to the drivers
- [ ] Has UI changes - Needs but not implemented yet
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates

Corresponding UI changes have not yet been implemented.

### CODE REVIEW

**Comment by jasonblais** [comment]
@harshilsharma Sorry for the delay on my part. Can you help with a rebase on the PR, and then @crspeller + @grundleborg can start review of the PR.

**Comment by harshilsharma63** [comment]
@jasonblais done.

**Comment by harshilsharma** [comment]
@jasonblais done.

File: store/sqlstore/upgrade.go [hunk 1]
```diff
@@ -514,3 +515,16 @@ func UpgradeDatabaseToVersion55(sqlStore SqlStore) {
...
 	// }
 }
+
+func UpgradeDatabaseToVersion56(sqlStore SqlStore) {
+	// TODO: Uncomment following condition when version 5.6.0 is released
+	//if shouldPerformUpgrade(sqlStore, VERSION_5_5_0, VERSION_5_6_0) {
+
+	// migrating user's accepted terms of service data into the new table
+	if sqlStore.DoesColumnExist("Users", "AcceptedTermsOfServiceId") {
+		sqlStore.GetMaster().Exec("INSERT INTO UserTermsOfService SELECT Id, AcceptedTermsOfServiceId as TermsOfServiceId, :CreateAt FROM Users WHERE AcceptedTermsOfServiceId != \"\" AND AcceptedTermsOfServiceId IS NOT NULL", map[string]interface{}{"CreateAt": model.GetMillis()})
```
Comment (crspeller) [inline store/sqlstore/upgrade.go]: Can you make sure this is acceptably fast on a DB with 60000 users?

Comment (harshilsharma) [reply store/sqlstore/upgrade.go]: I ran the query directly in the database and it took less than a second to run for a dummy dataset of 60000 users. Its a simple query for no joins and stuff, so it shouldn't be slow.

File: model/config.go [hunk 2]
```diff
@@ -1037,6 +1038,7 @@ type SupportSettings struct {
 	ReportAProblemLink          *string
 	SupportEmail                *string
 	CustomTermsOfServiceEnabled *bool
+	ReAcceptancePeriod          *int
```
Comment (jasonblais) [inline model/config.go]: Can we name this `CustomTermsOfServiceReAcceptancePeriod`, so it's consistent with other custom terms of service settings?

Comment (jasonblais) [reply model/config.go]: Please also help add to ../app/diagnostics.go

Comment (harshilsharma) [reply model/config.go]: 1. Renamed `ReAcceptancePeriod` to `CustomTermsOfServiceReAcceptancePeriod`.
1. Added `CustomTermsOfServiceReAcceptancePeriod` to app diagnostics.

**Comment by harshilsharma** [comment]
@lieut-data 

someone added the migration function to migrate to 5.6 and it has this code -

```
func UpgradeDatabaseToVersion56(sqlStore SqlStore) {
	// TODO: Uncomment following condition when version 5.5.0 is released
	//if shouldPerformUpgrade(sqlStore, VERSION_5_5_0, VERSION_5_6_0) {
	sqlStore.CreateColumnIfNotExists("PluginKeyValueStore", "ExpireAt", "bigint(20)", "bigint", "0")
	//	saveSchemaVersion(sqlStore, VERSION_5_5_0)
	//}
}
```

It seems that the version in first TODO, as well as in `saveSchemaVersion` is incorrect. Can you confirm this so I can rebase with master correctly before making the changes?

**Comment by lieut-data** [comment]
@harshilsharma, yes, that looks to be incorrect -- good catch. Did you want to patch it as part of rebasing?

**Comment by harshilsharma** [comment]
@lieut-data yes, doing that.

**Comment by harshilsharma** [comment]
@lieut-data done.

**Pull request merged**